### ISSUE TITLE
Make leakScope behavior more granular

### ISSUE THREAD

**Issue by Sinjhin** [canjs/can-component#52]
Based on the two different behaviors that the `leakScope` boolean controls:
http://canjs.github.io/canjs/doc/can-component.prototype.leakScope.html

Do we want to change the `leakScope` functionality to allow for individual control of each behavior?

For example, to have only ViewModel leak to the component template:
```js
Component.extend({
    tag: "my-component",
    leakScope: vm,
    ViewModel: { message: "Hello World!" },
    view: stache("{{message}}")
})
```
Or just outer scope to user content:
```js
Component.extend({
    tag: "my-component",
    leakScope: scope,
    ViewModel: { message: "Hello World!" },
    view: stache("{{message}}")
})
```
Or both:
```js
Component.extend({
    tag: "my-component",
    leakScope: all,
    ViewModel: { message: "Hello World!" },
    view: stache("{{message}}")
})
```

Alternative approach could be to replace `leakScope` with different props like:
```js
Component.extend({
    tag: "my-component",
    leakViewModel: true,
    leakOuterScope: true,
    ViewModel: { message: "Hello World!" },
    view: stache("{{message}}")
})
```

Thoughts?

**Comment by justinbmeyer** [comment]
Where would all come from? What would it be?  Did you intend for it to be a string?

**Comment by matthewp** [comment]
I think understanding/explaining the true/false options is hard enough, personally.

**Comment by justinbmeyer** [comment]
I think this makes it easier actually. There's two forms of leaking going on. People should be aware of that.

**Comment by marshallswain** [comment]
leakScope: true could be the same as "all"

Otherwise you pass "vm" or "scope".

**Comment by rjgotten** [comment]
> to have only ViewModel leak to the component template:
> Or just outer scope to user content:

These are both accomplished by setting `leakScope: false` and are not the 'leaked' data. The leaked data when `leakScope:true` is view model to user content and  outer scope to component template.

This nicely illustrates how badly understood `leakScope` is and just how much of a mess it is.
Imho, there's only one way to properly deal with this: remove it and adopt the `leakScope:false` behavior as fixed, so that component templates and viewmodels are always isolated and composition is always safe.

Then extend `<content>` or whatever slot-based replacement with a way to optionally pass a value which should be put on top of the scope stack while evaluating the user content. This is a more fine-grained; controlled; and local version of the view model to user content leak, which is practically the only useful direction of the scope leak. If you need to leak the other way, your application has serious architectural issues.

See also my original proposal https://github.com/canjs/canjs/issues/2429 which was funneled into https://github.com/canjs/can-component/issues/4 as part of slots.

**Comment by matthewp** [comment]
Is there a common example where leakScope is useful? I'm not questioning that it is, but it would be nice for the sake of trying to come up with solutions to have some firm common use cases.

Think of how a counter example is often used to demonstrate the advantage of event streams. Is there something like that for scope leaking?

**Comment by justinbmeyer** [comment]
It's very nice to leak scope into the `USER DOM`. 

can-connect/tag is an example of this:

```
<my-model get-list="{}">
  {{#each value}} ..... {{/each}}
</my-model>
``` 

I agree with @rjgotten that this should be part of `<content>` like : `<content add-scope=".">`


Regarding allowing `SHADOW DOM` to access parent `scope` ...

In general, it's lazy and makes things hard to test.  

Though it's useful when some components are rather tightly coupled.

It's used here to allow the `<h2>` to activate itself by calling the `makeActive()` of the parent VM: https://github.com/canjs/canjs/blob/455a50ca7b6cf8a104c0ce636a9fda2c25eb503d/demos/can-component/accordion.html

This was made before we could (more) easily pass methods from the parent to the child.  But with that, to use an `<accordion>` you'd have to either:

Always pass down the `makeActive` method: 

```
<accordion>
     <panel title="Contacts" {make-active}="@makeActive"> ... </panel>
     <panel title="Libraries" {make-active}="@makeActive"> ... </panel>
```

Or copy this in inserted:

```js
		inserted: function(){
			var vm = canViewModel(this.element.parentNode);
			vm.addPanel(this.viewModel);
                        this.viewModel.parentViewModel = vm;
		},
		removed: function(){
			this.viewModel.parentViewModel.removePanel(this.viewModel);
                        this.viewModel.parentViewModel = null;
		}
```


Wiring up this sort of `parent-child` relationship has long been indirectly requested, but I've not pursued it because I'd have to see a way to test it.

Perhaps some way of specifying parent or children and how they translate to properties or method calls:

```js
tag: "panel",
relationships: {
  ":parent" : "parentViewModel"
}

tag: "accordion"
relationships: {
  "panel" : "addPanel / removePanel"
}
```

Ideally, it should be declarative, not imperative like this.  But this really belongs in another issue.

**Comment by matthewp** [comment]
I think first-class templates are a better solution to the problems that leakScope tries to solve. What's nice about first-class templates is that they become part of a component's API (you expect your firstElementChild to be a template).

```handlebars
<my-model get-list="{}">
  <template>
    {{#each values}}
      <div>My model instance {{id}}</div>
    {{/each}}
  </template>
</my-model>
```

Here the `my-model` component creates a renderer from the child template and pass `values` (and maybe some other properties) to render it.

Instead of leaking the entire scope, the parent component is in control of what properties it passes to the child template.

With stache this is less nice since it needs to use a script tag:

```handlebars
<my-model get-list="{}">
  <script type="text/stache">
    {{#each values}}
      <div>My model instance {{id}}</div>
    {{/each}}
  </script>
</my-model>
```

Of course this can work today, you only need to write your components to do this. And you don't have to think about / worry about leakScope.

**Comment by justinbmeyer** [comment]
What are first class templates?

**Comment by justinbmeyer** [comment]
The native template won't work (if this is what you mean) as not all stache syntaxes are valid.

**Comment by justinbmeyer** [comment]
I'm actually a bit confused by what you are proposing and why it solves leak scope. It seems more like you are taking about <slot> stuff.

**Comment by matthewp** [comment]
> What are first class templates?

It's just a template, but one that a component *expects*. Similar to how bit-tabs expects bit-panels as children, some components would expect a template as a child

> I'm actually a bit confused by what you are proposing and why it solves leak scope. It seems more like you are taking about <slot> stuff.

No, this has nothing to do with slot. This is about replacing the need for `leakScope: true` with using templates instead. It has the benefits of:

1. The component doesn't leak its entire scope, but rather renders the child template with just the properties it needs.
2. There's an explicit contract for the component.
3. It's a more generic solution (devs already know about templates and how to use them).

> The native template won't work (if this is what you mean) as not all stache syntaxes are valid.

I addressed that in my comment.

**Comment by rjgotten** [comment]
@matthewp 

I.e. `<content>` refers to a particular 'strongly typed' constraint on the template, which defines what input parameters the template will be passed. I think that sounds a lot like what I proposed with `<content can-scope="..." />`, albeit with a much nicer API.

Thinking out loud here: would that mean 'first class' templates are implemented by taking user-content and upgrading it into some kind of bespoke component with data bindings to propagate parameters into a scope/viewmodel that will live at the top of the scope stack in the user-content?

I.e. you'd implement a looped template somewhat like:

```
{{#each items}}
  <content {item}="." {index}="%index" />
{{/each}}
```

If so: awesome. That looks like an incredibly clean way of handling things. Even better if this comes with the full package of supporting two-way databinding; passing methods; etc. It should make inter-communicating components (such as aforementioned accordion and panel) also feel a lot more structured and less 'hacked'.

**Comment by matthewp** [comment]
@rjgotten Yeah, something like that. I don't know if an element or a scope function is more appropriate (thinking something like `{{subtemplate(foo=bar baz='qux')}}`, but I probably wouldn't choose `<content/>` because of its legacy meaning.

Nevertheless, as discussed above, nested templates aren't possible in Stache b/c it uses `<script>`. So I guess leakScope has to remain to handle script-based templating languages.
